tests/test_prepare_annotations.py

```diff
@@ -100,3 +103,146 @@ def __prepare_pydantic_annotations__(
 
     a = TypeAdapter(Annotated[MyType, Lt(420)])
     assert a.core_schema == de.IsPartialDict(core_schema.int_schema(gt=123, lt=420))
+
+
+roman_numbers = {
+    'I': 1,
+    'II': 2,
+    'III': 3,
+    'IV': 4,
+    'V': 5,
+    'VI': 6,
+    'VII': 7,
+    'VIII': 8,
+    'IX': 9,
+    'X': 10,
+}
+roman_numbers_lookup = {v: k for k, v in roman_numbers.items()}
+
+
+class Roman:
+    def __init__(self, value: int, *, allow_nan: bool = False) -> None:
+        try:
+            self.numeral = roman_numbers_lookup[value]
+        except KeyError:
+            if allow_nan:
+                self.numeral = 'NAN'
+            else:
+                raise ValueError('not a roman numeral')
+
+    @classmethod
+    def from_str(cls, value: str, *, allow_nan: bool = False) -> 'Roman':
+        if value == 'NAN':
+            return cls(0, allow_nan=allow_nan)
+
+        try:
+            value = roman_numbers[value]
+        except KeyError:
+            raise ValueError('invalid input')
+        else:
+            return cls(value)
+
+    @property
+    def int(self) -> int:
+        try:
+            return roman_numbers[self.numeral]
+        except KeyError:
+            raise ValueError('not a numeral')
+
+    def __add__(self, other):
+        return Roman(self.int + other.int)
+
+    def __sub__(self, other):
+        return Roman(self.int - other.int)
+
+    def __eq__(self, other):
+        if isinstance(other, Roman):
+            return self.numeral == other.numeral
+        else:
+            return False
+
+    def __str__(self):
+        return self.numeral
+
+    def __repr__(self):
+        return f'Roman({self.numeral})'
+
+    @classmethod
+    def __prepare_pydantic_annotations__(
+        cls, source_type: Any, annotations: Sequence[Any], config: ConfigDict
+    ) -> Tuple[Any, List[Any]]:
+        allow_inf_nan = config.get('allow_inf_nan', False)
+        for an in annotations:
+            if isinstance(an, AllowInfNan):
+                allow_inf_nan = an.allow_inf_nan
+                break
+
+        return source_type, [RomanValidator(allow_inf_nan)]
+
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source_type: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
```

Comment:
@adriangb  my point is that we could not require a `__get_pydantic_core_schema__` if `__prepare_pydantic_annotations__` is defined since, like this case, it could do nothing.

Probably not worth doing right now.

Reply:
I'm not sure I'm understanding what you want. Does https://github.com/pydantic/pydantic/pull/5886 solve this?

Reply:
I guess so I'll try it.